jormungandr/src/blockchain/reference_cache.rs

```diff
@@ -124,7 +124,7 @@ impl RefCacheData {
     pub fn poll_purge(&mut self) -> Poll<(), timer::Error> {
         loop {
             match self.expirations.poll()? {
-                Async::NotReady => return Ok(Async::Ready(())),
+                Async::NotReady => return Ok(Async::NotReady),
```

Comment:
no it is a `Ready(())` that you want to return, the reason being that otherwise the cache will never free from the purging and will wait to have more expired values.

In the case of **purging** the expiration queue you want to unblock the purging task as soon as you have had all the elements who needs to be removed. `NotReady` is returned as soon as there are still elements in the expiration poll that haven't expired. So no, the previous value was correct.

Reply:
It's a big no-no to return `Ready` when all sources (in this context) polled `NotReady`.
This way, the re-poll of this task has been scheduled in the reactor to when the next expiration comes up, but returning `Ready` would leed to it being polled again to busy loop. I had the same problem with client code.

Reply:
Theoretically I would say yes. But the way it is used in this case is not the same. We are polling all elements that are ready to be removed. Not polling the element that we want to keep. And this is what this is doing here. If you return `NotReady` the task will wait for **all** the element in the expirations to be polled until there is none. The issue goes back to #703 where I have made the same assumption you are making about how it works here.

Reply:
I thought the `DelayQueue` works just like all the other async objects: if the poll returns `NotReady`, it has nothing to yield now, but has scheduled a wakeup for when it might have more. This is what the docs on the `DelayQueue` say. So re-polling before the scheduled time makes no sense. There is one edge case when a new item is inserted and its expiration time is shorter than a previously scheduled wakeup, but I'd suppose [it is handled](https://docs.rs/tokio-timer/0.2.11/src/tokio_timer/delay_queue.rs.html#411) in the implementation?

Reply:
yes but that's not how it is used, we poll every `N` seconds to `purge` the expired objects. This is needed because we cannot keep hold of the object all the time. Otherwise we cannot add new entries in the delay queue. Going back to the deadlock fixed by #703

Reply:
One thing is that we need to sit down and convert all these tasks to run properly async, without `wait` calls in the message handlers. Then these polls will work as they are intended to and returning `Ready()` on non-ready queue will busyloop. Even now, I believe polling every N seconds will yield the expired entry; where did the deadlock come from?

Reply:
Removed this change for now.